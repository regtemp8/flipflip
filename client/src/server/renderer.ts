import createCache, { type EmotionCache } from '@emotion/cache'
import { configureStore } from '@reduxjs/toolkit'
import { AppStorage, SystemConstants } from 'flipflip-common'
import createAppReducer from '../store/app/slice'
import createAudioReducer from '../store/audio/slice'
import createCaptionScriptReducer from '../store/captionScript/slice'
import createClipReducer from '../store/clip/slice'
import createLibrarySourceReducer from '../store/librarySource/slice'
import createPlaylistReducer from '../store/playlist/slice'
import createSceneReducer from '../store/scene/slice'
import createSceneGroupReducer from '../store/sceneGroup/slice'
import createTagReducer from '../store/tag/slice'
import playersSlice from '../store/player/slice'
import videoClipperSlice from '../store/videoClipper/slice'
import scenePickerSlice from '../store/scenePicker/slice'
import captionScriptorSlice from '../store/captionScriptor/slice'
import sceneDetailSlice from '../store/sceneDetail/slice'
import audioLibrarySlice from '../store/audioLibrary/slice'
import sourceScraperSlice from '../store/sourceScraper/slice'
import createConstantsReducer from '../store/constants/slice'
import { fromAppStorage } from '../store/app/convert'
import createDisplayReducer from '../store/display/slice'
import createDisplayViewReducer from '../store/displayView/slice'
import createScenePlaylistItemReducer from '../store/scenePlaylistItem/slice'
import createDisplayPlaylistItemReducer from '../store/displayPlaylistItem/slice'

export function createEmotionCache(): EmotionCache {
  return createCache({ key: 'css' })
}

export function createReduxStore(
  appStorage?: AppStorage,
  constants?: SystemConstants
) {
  const state = appStorage != null ? fromAppStorage(appStorage) : undefined
  return configureStore({
    reducer: {
      app: createAppReducer(state?.app),
      audio: createAudioReducer(state?.audio),
      captionScript: createCaptionScriptReducer(state?.captionScript),
      clip: createClipReducer(state?.clip),
      librarySource: createLibrarySourceReducer(state?.librarySource),
      playlist: createPlaylistReducer(state?.playlist),
      displayPlaylistItem: createDisplayPlaylistItemReducer(
        state?.displayPlaylistItem
      ),
      scenePlaylistItem: createScenePlaylistItemReducer(
        state?.scenePlaylistItem
      ),
      scene: createSceneReducer(state?.scene),
      sceneGroup: createSceneGroupReducer(state?.sceneGroup),
      display: createDisplayReducer(state?.display),
      displayView: createDisplayViewReducer(state?.displayView),
      tag: createTagReducer(state?.tag),
      constants: createConstantsReducer(constants),
      // components
      players: playersSlice,
      videoClipper: videoClipperSlice,
      scenePicker: scenePickerSlice,
      captionScriptor: captionScriptorSlice,
      sceneDetail: sceneDetailSlice,
      audioLibrary: audioLibrarySlice,
      sourceScraper: sourceScraperSlice
    }
  })
}
