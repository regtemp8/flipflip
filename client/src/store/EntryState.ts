export interface Identifiable {
  id: number
}

export interface EntryState<T extends Identifiable> {
  name: string
  nextID: number
  entries: Record<number, T>
}

export interface EntryUpdate<T> {
  id: number
  value: T
}

export function isEntry<T extends Identifiable>(
  state: EntryState<T>,
  id: number
) {
  return state.entries[id] != null
}

export function setNextID<T extends Identifiable>(
  state: EntryState<T>,
  entry: T
) {
  if (state.nextID <= entry.id) {
    state.nextID = entry.id + 1
  }
}

export function getEntry<T extends Identifiable>(
  state: EntryState<T>,
  id: number
): T {
  return state.entries[id]
}

export function setEntry<T extends Identifiable>(
  state: EntryState<T>,
  entry: T
) {
  state.entries[entry.id] = entry
  setNextID(state, entry)
}

export function createEntry<T extends Identifiable>(
  state: EntryState<T>,
  entry: T
) {
  if (!isEntry(state, entry.id)) {
    state.entries[entry.id] = entry
    setNextID(state, entry)
  } else {
    throw new Error(
      `${state.name}: Create failed. Entry with ID ${entry.id} already exists.`
    )
  }
}

export function createEntries<T extends Identifiable>(
  state: EntryState<T>,
  entries: T[]
) {
  for (const entry of entries) {
    if (isEntry(state, entry.id)) {
      throw new Error(
        `${state.name}: Create failed. Entry with ID ${entry.id} already exists.`
      )
    }
  }

  entries.forEach((s) => (state.entries[s.id] = s))
  const maxID = Math.max(...entries.map((s) => s.id))
  setNextID(state, { id: maxID })
}

export function updateEntry<T extends Identifiable>(
  state: EntryState<T>,
  entry: T
) {
  if (isEntry(state, entry.id)) {
    state.entries[entry.id] = entry
  } else {
    throw new Error(
      `${state.name}: Update failed. Entry with ID ${entry.id} doesn't exist.`
    )
  }
}

export function deleteEntry<T extends Identifiable>(
  state: EntryState<T>,
  id: number
) {
  if (isEntry(state, id)) {
    delete state.entries[id]
  } else {
    throw new Error(
      `${state.name}: Delete failed. Entry with ID ${id} doesn't exist.`
    )
  }
}
