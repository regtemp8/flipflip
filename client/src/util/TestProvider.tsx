import * as React from 'react'
import { Provider } from 'react-redux'
import { createTheme } from '@mui/material'
import { StyledEngineProvider, ThemeProvider } from '@mui/material/styles'
import defaultTheme from 'flipflip-common'

export default class TestProvider extends React.Component {
  readonly props: {
    children: any
    store: any
  }

  disableTransitions = {
    defaultProps: {
      disablePortal: true,
      hideBackdrop: true,
      TransitionComponent: ({ children }) => children
    }
  }

  render() {
    const theme = createTheme({
      ...defaultTheme,
      // disable Transitions when runing tests, throws errors otherwise.
      components: {
        MuiMenu: this.disableTransitions,
        MuiPopover: this.disableTransitions,
        MuiDialog: this.disableTransitions,
        MuiModal: this.disableTransitions
      }
    } as any)
    return (
      <Provider store={this.props.store}>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>{this.props.children}</ThemeProvider>
        </StyledEngineProvider>
      </Provider>
    )
  }
}
