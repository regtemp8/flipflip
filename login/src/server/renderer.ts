import createCache, { type EmotionCache } from "@emotion/cache";

export function createEmotionCache(nonce?: string): EmotionCache {
  return createCache({ key: "css", nonce });
}
