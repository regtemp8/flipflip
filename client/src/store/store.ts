/// <reference path="../global.d.ts"/>
import { createReduxStore } from '../server/renderer'

const store = createReduxStore(
  window.flipflipAppStorage,
  window.flipflipConstants
)

export default store
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
