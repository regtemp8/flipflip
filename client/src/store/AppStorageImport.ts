import { initialApp } from "./app/data/App";
import { initialAudioState } from "./audio/slice";
import { initialCaptionScriptState } from "./captionScript/slice";
import { initialClipState } from "./clip/slice";
import { initialLibrarySourceState } from "./librarySource/slice";
import { initialOverlayState } from "./overlay/slice";
import { initialPlaylistState } from "./playlist/slice";
import { initialSceneState } from "./scene/slice";
import { initialSceneGridState } from "./sceneGrid/slice";
import { initialSceneGroupState } from "./sceneGroup/slice";
import { initialTagState } from "./tag/slice";
import { EntryState } from "./EntryState";
import type Tag from "./tag/Tag";
import type SceneGroup from "./sceneGroup/SceneGroup";
import type SceneGrid from "./sceneGrid/SceneGrid";
import type Scene from "./scene/Scene";
import type Playlist from "./playlist/Playlist";
import type Overlay from "./overlay/Overlay";
import type LibrarySource from "./librarySource/LibrarySource";
import type Clip from "./clip/Clip";
import type CaptionScript from "./captionScript/CaptionScript";
import type Audio from "./audio/Audio";
import type App from "./app/data/App";

export interface AppStorageImport {
  app: App;
  audio: EntryState<Audio>;
  captionScript: EntryState<CaptionScript>;
  clip: EntryState<Clip>;
  librarySource: EntryState<LibrarySource>;
  overlay: EntryState<Overlay>;
  playlist: EntryState<Playlist>;
  scene: EntryState<Scene>;
  sceneGrid: EntryState<SceneGrid>;
  sceneGroup: EntryState<SceneGroup>;
  tag: EntryState<Tag>;
}

export const initialAppStorageImport: AppStorageImport = {
  app: initialApp,
  audio: initialAudioState,
  captionScript: initialCaptionScriptState,
  clip: initialClipState,
  librarySource: initialLibrarySourceState,
  overlay: initialOverlayState,
  playlist: initialPlaylistState,
  scene: initialSceneState,
  sceneGrid: initialSceneGridState,
  sceneGroup: initialSceneGroupState,
  tag: initialTagState,
}