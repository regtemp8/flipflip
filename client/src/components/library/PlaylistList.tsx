import React, { useState } from 'react'
import { cx } from '@emotion/css'

import {
  Card,
  CardContent,
  CardMedia,
  Grid,
  type Theme,
  Typography
} from '@mui/material'

import { makeStyles } from 'tss-react/mui'

import AudiotrackIcon from '@mui/icons-material/Audiotrack'
import { useAppSelector } from '../../store/hooks'
import { selectAppPlaylistThumbs } from '../../store/playlist/selectors'

const useStyles = makeStyles()((theme: Theme) => ({
  emptyMessage: {
    textAlign: 'center',
    marginTop: '25%'
  },
  emptyMessage2: {
    textAlign: 'center'
  },
  media: {
    height: 0,
    paddingTop: '100%' // 16:9  = 56.25%
  },
  mediaIcon: {
    width: '100%',
    height: 'auto'
  },
  underlineTitle: {
    textDecoration: 'underline'
  },
  pointer: {
    cursor: 'pointer'
  },
  root: {
    borderRadius: 1
  },
  cardContent: {
    '&:last-child': {
      paddingBottom: theme.spacing(2)
    }
  },
  artist: {
    '&:hover': {
      textDecoration: 'underline'
    }
  }
}))

export interface PlaylistListProps {
  showHelp: boolean
  onClickPlaylist: (playlist: string) => void
}

function PlaylistList(props: PlaylistListProps) {
  const playlistThumbs = useAppSelector(selectAppPlaylistThumbs())
  const [hover, setHover] = useState<any>()

  const onMouseEnter = (album: string) => {
    setHover(album)
  }

  const onMouseLeave = () => {
    setHover(null)
  }

  const { classes } = useStyles()
  if (Object.keys(playlistThumbs).length === 0) {
    return (
      <React.Fragment>
        <Typography
          component="h1"
          variant="h3"
          color="inherit"
          noWrap
          className={classes.emptyMessage}
        >
          乁( ◔ ౪◔)「
        </Typography>
        <Typography
          component="h1"
          variant="h4"
          color="inherit"
          noWrap
          className={classes.emptyMessage2}
        >
          Nothing here
        </Typography>
        {props.showHelp && (
          <Typography
            component="h1"
            variant="h6"
            color="inherit"
            noWrap
            className={classes.emptyMessage2}
          >
            Create playlists by clicking "Add to Playlist" in the sidebar
          </Typography>
        )}
      </React.Fragment>
    )
  }

  return (
    <Grid container spacing={2}>
      {Object.keys(playlistThumbs).map((p) => {
        const thumbs = playlistThumbs[p].map((thumb) =>
          thumb.replace(/\\/g, '/')
        )
        if (thumbs.length === 2) {
          thumbs.push('')
          thumbs.push('')
          thumbs[3] = thumbs[1]
          thumbs[1] = ''
        }
        return (
          <Grid
            key={p}
            item
            xs={6}
            sm={4}
            md={3}
            lg={2}
            className={classes.pointer}
            onClick={() => props.onClickPlaylist(p)}
            onMouseEnter={() => onMouseEnter(p)}
            onMouseLeave={onMouseLeave}
          >
            <Card classes={{ root: classes.root }}>
              <Grid container>
                {thumbs.map((t, index) => (
                  <Grid item xs={thumbs.length === 1 ? 12 : 6} key={index}>
                    {t && <CardMedia className={classes.media} image={t} />}
                  </Grid>
                ))}
                {thumbs.length === 0 && (
                  <Grid item xs={12}>
                    <AudiotrackIcon className={classes.mediaIcon} />
                  </Grid>
                )}
              </Grid>
              <CardContent classes={{ root: classes.cardContent }}>
                <Typography
                  className={cx(hover === p && classes.underlineTitle)}
                  noWrap
                  variant="body1"
                >
                  {p}
                </Typography>
              </CardContent>
            </Card>
          </Grid>
        )
      })}
    </Grid>
  )
}

;(PlaylistList as any).displayName = 'PlaylistList'
export default PlaylistList
