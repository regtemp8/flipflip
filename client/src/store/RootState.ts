import { RootState } from "./store";
import { initialApp } from "./app/data/App";
import { initialAudioState } from "./audio/slice";
import { initialCaptionScriptState } from "./captionScript/slice";
import { initialClipState } from "./clip/slice";
import { initialSystemConstants } from "./constants/SystemConstants";
import { initialLibrarySourceState } from "./librarySource/slice";
import { initialOverlayState } from "./overlay/slice";
import { initialPlaylistState } from "./playlist/slice";
import { initialSceneState } from "./scene/slice";
import { initialSceneGridState } from "./sceneGrid/slice";
import { initialSceneGroupState } from "./sceneGroup/slice";
import { initialTagState } from "./tag/slice";
import { initialPlayersState } from "./player/slice";
import { initialVideoClipperState } from "./videoClipper/slice";
import { initialScenePickerState } from "./scenePicker/slice";
import { initialCaptionScriptorState } from "./captionScriptor/slice";
import { initialSceneDetailState } from "./sceneDetail/slice";
import { initialAudioLibraryState } from "./audioLibrary/slice";
import { initialSourceScraperState } from "./sourceScraper/slice";

export const initialRootState: RootState = {
  app: initialApp,
  audio: initialAudioState,
  captionScript: initialCaptionScriptState,
  clip: initialClipState,
  constants: initialSystemConstants,
  librarySource: initialLibrarySourceState,
  overlay: initialOverlayState,
  playlist: initialPlaylistState,
  scene: initialSceneState,
  sceneGrid: initialSceneGridState,
  sceneGroup: initialSceneGroupState,
  tag: initialTagState,
  players: initialPlayersState,
  videoClipper: initialVideoClipperState,
  scenePicker: initialScenePickerState,
  captionScriptor: initialCaptionScriptorState,
  sceneDetail: initialSceneDetailState,
  audioLibrary: initialAudioLibraryState,
  sourceScraper: initialSourceScraperState
}