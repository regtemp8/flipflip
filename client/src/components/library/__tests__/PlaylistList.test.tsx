import React from 'react'
import { describe, it, expect } from '@jest/globals'
import renderer from 'react-test-renderer'
import PlaylistList from '../PlaylistList'
import TestProvider from '../../../util/TestProvider'
import store from '../../../store/store'

describe('PlaylistList', () => {
  it('should match snapshot', () => {
    const component = renderer.create(
      <TestProvider store={store}>
        <PlaylistList showHelp={false} onClickPlaylist={(playlist) => {}} />
      </TestProvider>
    )

    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
