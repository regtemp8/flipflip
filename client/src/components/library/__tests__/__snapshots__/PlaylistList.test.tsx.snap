// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaylistList should match snapshot 1`] = `
Array [
  <h1
    className="MuiTypography-root MuiTypography-h3 MuiTypography-noWrap PlaylistList-emptyMessage-1 css-1yk4hom-MuiTypography-root"
  >
    乁( ◔ ౪◔)「
  </h1>,
  <h1
    className="MuiTypography-root MuiTypography-h4 MuiTypography-noWrap PlaylistList-emptyMessage2-2 css-1r3v6n7-MuiTypography-root"
  >
    Nothing here
  </h1>,
]
`;
